File: studio/web-app/src/app/api-bak/core-model.ts
// direct import to avoid require all common in core
// we depend on mylife-home-core only for this
// export * from 'mylife-home-core/dist/store/model';

export const enum StoreItemType {
  COMPONENT = 'component',
  BINDING = 'binding'
}

export interface StoreItem {
  readonly type: StoreItemType;
  readonly config: ComponentConfig | BindingConfig;
}

export interface ComponentConfig {
  readonly id: string;
  readonly plugin: string;
  readonly config: { [name: string]: any };
}

export interface BindingConfig {
  readonly sourceComponent: string;
  readonly sourceState: string;
  readonly targetComponent: string;
  readonly targetAction: string;
}

File: studio/web-api/src/core_model.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::register_ts;

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "core-model.ts")]
pub enum StoreItemType {
    Component,
    Binding,
}

register_ts!(StoreItemType);

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "core-model.ts")]
pub struct StoreItem {
    pub r#type: StoreItemType,
    #[ts(type = "any")]
    pub config: serde_json::Value,
}

register_ts!(StoreItem);

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "core-model.ts")]
pub struct ComponentConfig {
    pub id: String,
    pub plugin: String,
    #[ts(type = "{ [key: string]: any }")]
    pub config: std::collections::HashMap<String, serde_json::Value>,
}

register_ts!(ComponentConfig);

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export_to = "core-model.ts")]
pub struct BindingConfig {
    pub source_component: String,
    pub source_state: String,
    pub target_component: String,
    pub target_action: String,
}

register_ts!(BindingConfig);
